PParamsUpdate enactment fix

lehins opened the pull request:
# Description

Fixes #3835 

Original implementation was buggy and did not produced any groups for PParam groups, which resulted in DRep thresholds to be set to zero for PParamUpdates, yielding all PParam proposals be enacted, regardless on what DReps though about them :smile: 

One more fix included is a PParamUpdate wellformdness check that ensures empty PParams can't be proposed.  

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

lehins committed:
Add a property test that checks that one of the trhesholds is always selected

lehins committed:
Reimplement DRep threshold computation

This is an alternative approach to figuring out thresholds. Original
approach was buggy and did not work correctly, which resulted in #3835

Bring back the module

Bring back the module

Soupstraw commented on eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs:
```diff
@@ -493,155 +538,160 @@ conwayPParamsPairs pp =
-    , cppA0 = minBound
-    , cppRho = minBound
-    , cppTau = minBound
+    { cppMinFeeA = THKD (Coin 0)
```
Lots of redundant parentheses here

lehins replied:
Thank you for pointing it out. Result of search and replace. I'll fix it

lehins replied:
Removed

Soupstraw commented on libs/cardano-ledger-core/src/Cardano/Ledger/Core/PParams.hs:
```diff
@@ -196,7 +196,10 @@ instance Updatable (U1 a) (U1 u) where
+  Updatable ((a1 :*: a2) a) ((u1 :*: u2) u)
+  where
   applyUpdate (x1 :*: x2) (u1 :*: u2) = applyUpdate x1 u1 :*: applyUpdate x2 u2
 
```
The Generic updating should work with `THKD` if you add an instance for `Updatable (K1 t x a) (K1 t (THKD f x u))` where the `applyUpdate` just unwraps the `THKD` and calls `applyUpdate` again

lehins replied:
I have tried it, but couldn't figure it out. My generics Kung fu isn't strong enough.

Soupstraw commented on eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs:
```diff
@@ -94,70 +166,70 @@ import Numeric.Natural (Natural)
 -- * @dRepDeposit@
 -- * @dRepActivity@
 data ConwayPParams f era = ConwayPParams
-  { cppMinFeeA :: !(HKD f Coin)
```
I like this approach

lehins replied:
Nice! :slightly_smiling_face:

teodanciu commented on eras/conway/impl/CHANGELOG.md:
```diff
@@ -1,9 +1,14 @@
+* Expose `pparamsUpdateThreshold`
+* Fix [#3835](https://github.com/input-output-hk/cardano-ledger/issues/3835)
+* Rename `PParamGroup` to `PPGroup` and `GovernanceGroup` to `GovGroup`
+* Intorduce `THKD` and use it for `ConwayPParams`
```
Little typo here, "Introduce"

lehins replied:
Fixed

teodanciu commented on libs/cardano-ledger-core/src/Cardano/Ledger/Ap.hs:
```diff
@@ -1,115 +0,0 @@
-{-# LANGUAGE GADTs #-}
```
Shouldn't this removal be reflected in a version change and changelog of cardano-ledger-core, since we were exposing it from the library?

lehins replied:
Yeah, you are right. Thank you. I'll fix that.

lehins replied:
I decided to instead deprecate the module and will remove it once we have bunch of other breaking changes :slightly_smiling_face:

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs:
```diff
@@ -380,6 +416,7 @@ instance Crypto c => ConwayEraPParams (ConwayEra c) where
         isValid (/= zero) ppuPoolDepositL
       , isValid (/= zero) ppuGovActionDepositL
       , isValid (/= zero) ppuDRepDepositL
+      , ppu /= emptyPParamsUpdate
```
but I would put this in a different commit or maybe at least mention it in the commit message, in case we have to track  it down (for example for changelogs).  But it's up to you of course

lehins replied:
Very good point. I've put it into a separate commit

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs:
```diff
@@ -868,26 +895,158 @@ ppDRepDepositL = ppLens . hkdDRepDepositL @era @Identity
+  PParams era
+ppUpdateCostModels ppu pp =
+  case ppu ^. ppuCostModelsL of
+    SNothing -> pp
```
How come this pparam gets special treatment if it's empty, as opposed to other ppu lenses of type StrictMaybe?

lehins replied:
Good question. Here is why: #3631 and the fix #3771

teodanciu replied:
Great, I see now the connection, thanks :+1:

teodanciu reviewed (approved):
Left a few comments and questions.  
Looks good to me.

lehins committed:
Prevent empty PParamUpdate from being proposed.

This is very important, otherwise we can't figure out the group and thus the
threshold for DReps, making a proposla trivially enactable, with the CC
support that is.

Related change to the spec:
https://github.com/input-output-hk/formal-ledger-specifications/pull/274

lehins committed:
Rewrite `conwayApplyPPUpdates` to use record syntax

GHC-8.10 has a bug where current setup of type families and usage of
type class functions to define standalone functions that depnend on the
same type class results in compiler going into infinite loop.

GHC bug report: https://gitlab.haskell.org/ghc/ghc/-/issues/21973

lehins committed:
f
